diff --git a/_autosetup b/_autosetup
index 636a4eb56b..ec1b66f0b4 100755
--- a/_autosetup
+++ b/_autosetup
@@ -19,7 +19,7 @@ esac
 check_version()
 {
     dir=`pwd`
-    cd /tmp
+    cd "${TMPDIR:-/tmp}"
     foundit=
     ## get current version of $1
     desired=`echo $2 | awk -F. '{print $1*100+$2}'`
diff --git a/client/Makefile.am b/client/Makefile.am
index a85c8c1891..68179a9ccc 100644
--- a/client/Makefile.am
+++ b/client/Makefile.am
@@ -25,7 +25,7 @@ if OS_DARWIN
 endif
 
 bin_PROGRAMS = boinc_client boinccmd
-if !OS_WIN32
+if OS_DARWIN
 bin_PROGRAMS += switcher
 endif
 
diff --git a/client/hostinfo_unix.cpp b/client/hostinfo_unix.cpp
index 2cf9af3a18..227d86f599 100644
--- a/client/hostinfo_unix.cpp
+++ b/client/hostinfo_unix.cpp
@@ -1414,11 +1414,11 @@ int HOST_INFO::get_os_info() {
 #if HAVE_SYS_UTSNAME_H
     struct utsname u;
     uname(&u);
-#ifdef ANDROID
+#ifdef __ANDROID__
     safe_strcpy(os_name, "Android");
 #else
     safe_strcpy(os_name, u.sysname);
-#endif //ANDROID
+#endif // __ANDROID__
 #if defined(__EMX__) // OS2: version is in u.version
     safe_strcpy(os_version, u.version);
 #elif defined(__HAIKU__)
diff --git a/lib/synch.cpp b/lib/synch.cpp
index 60124de8c1..7a0acbe833 100644
--- a/lib/synch.cpp
+++ b/lib/synch.cpp
@@ -18,7 +18,7 @@
 
 // interfaces for accessing sempahores
 
-#ifndef ANDROID
+#ifndef __ANDROID__
 
 #include "config.h"
 #include <cstdio>
